mlir/lib/Dialect/MIOpen/Tuning/solver/conv_hip_implicit_gemm_fwd_v4r4.cpp

```diff
@@ -0,0 +1,506 @@
+      CalculateGemmABlockCopyPerformanceParameters(ctx);
+
+  if (failed(res))
+    return failure();
+
+  // check blockwise copy of B matrix
+  std::tie(std::ignore, std::ignore, std::ignore, std::ignore, res) =
+      CalculateGemmBBlockCopyPerformanceParameters(ctx);
+
+  if (failed(res))
+    return failure();
+
+  // check threadwise copy of C matrix
+  std::tie(std::ignore, res) =
+      CalculateGemmCThreadCopyPerformanceParameters(ctx);
+
+  if (failed(res))
+    return failure();
+
+  // check LDS allocation
+  std::size_t lds_size = 0;
+  std::tie(lds_size, res) = CalculateLdsNumberOfByte(ctx);
+
+  if (succeeded(res) && lds_size <= get_lds_max_number_of_byte())
+    return success();
+
+  return failure();
+}
+
+LogicalResult
+PerformanceImplicitGemmV4R4Fwd::EuristicInit(const ConvolutionContext &ctx) {
+  PerformanceImplicitGemmV4R4Fwd config;
+
+  using ConfigParam = llvm::SmallVector<int64_t, 6>;
+  // blockSize M/block N/block K/block M/thread N/thread
+  llvm::SmallVector<ConfigParam, 21> initParams = {
+      // clang-format off
+      {256, 128, 128, 8, 4, 4},
+      {128, 128, 64, 8, 4, 4},
+      {128, 64, 128, 4, 4, 4},
+      {64, 64, 64, 16, 4, 4},
+      {64, 32, 64, 16, 2, 4},
+      {64, 64, 32, 16, 4, 2},
+      {64, 32, 32, 4, 2, 2}};
+  // clang-format on
+  /* MIOpen logic */
+  /*
+        {256, 128, 128, 16, 4, 4}, {256, 128, 128, 8, 4, 4},
+        {256, 128, 128, 4, 4, 4},  {128, 128, 64, 16, 4, 4},
+        {128, 128, 64, 8, 4, 4},   {128, 128, 64, 4, 4, 4},
+        {128, 64, 128, 16, 4, 4},  {128, 64, 128, 8, 4, 4},
+        {128, 64, 128, 4, 4, 4},   {64, 64, 64, 16, 4, 4},
+        {64, 64, 64, 8, 4, 4},     {64, 64, 64, 4, 4, 4},
+        {64, 64, 32, 16, 4, 2},    {64, 64, 32, 8, 4, 2},
+        {64, 64, 32, 4, 4, 2},     {64, 32, 64, 16, 2, 4},
+        {64, 32, 64, 8, 2, 4},     {64, 32, 64, 4, 2, 4},
+        {64, 32, 32, 16, 2, 2},    {64, 32, 32, 8, 2, 2},
+        {64, 32, 32, 4, 2, 2}};
+  */
+  for (auto &param : initParams) {
+    config = {param[0], param[1], param[2], param[3], param[4], param[5]};
+    if (succeeded(config.IsValid(ctx))) {
+
+      LLVM_DEBUG(llvm::dbgs() << "Successfully picked tuning params from backup"
+                              << " path.\n");
+      *this = config;
+      return success();
+    }
+  }
+
+  llvm::errs() << "FATAL ERROR! COULD NOT FIND VALID TUNING PARAMETERS!\n";
+  return failure();
+}
+
+std::tuple<int64_t, int64_t, int64_t>
+ConvHipImplicitGemmV4R4Fwd::CalculateGemmSize(const ConvolutionContext &ctx) {
+  auto dimIndexVal = ctx.dimIndexVal;
+  const auto n = dimIndexVal["no"].second;
+  const auto k = dimIndexVal["k"].second;
+  const auto c = dimIndexVal["c"].second;
+  const auto ho = dimIndexVal["ho"].second;
+  const auto wo = dimIndexVal["wo"].second;
+  const auto y = dimIndexVal["y"].second;
+  const auto x = dimIndexVal["x"].second;
+  llvm::errs() << "n=" << n << " ho=" << ho << " wo=" << wo << "\n";
```

Comment:
Use `LLVM_DEBUG`. `llvm::errs()` will be sent to stderr, and can confuse users with real error messages.

This applies to multiple locations where you aim to output a debug message.